src/ExternalWalletStorage.sol

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.23;
+
+/// @title External Wallet Storage
+/// @notice Singleton contract for storing wallet state, protected by caller address and code hash
+contract ExternalWalletStorage {
+    /// @notice Storage layout used by each wallet instance
+    struct WalletStorage {
+        uint256 nextOwnerIndex;
+        uint256 removedOwnersCount;
+        mapping(uint256 index => bytes owner) ownerAtIndex;
+        mapping(bytes bytes_ => bool isOwner_) isOwner;
+    }
+
+    /// @notice Maps (wallet address, wallet code hash) to wallet storage
+    mapping(address => mapping(bytes32 => WalletStorage)) private walletStorage;
```

Comment:
nice idea. but needs a small fix to be accepted by the bundler's validation rules:
In order to be associated with the account's address, the last mapping has to be the address, not the first.
so change the mapping to 
`mapping(bytes32 => mapping(address => WalletStorage))`

another issue is allow an account to have arbitrary data stored in the external storage, but that is a separate issue.